src/Analyzers/Core/Analyzers/UseCoalesceExpression/AbstractUseCoalesceExpressionForIfNullCheckDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,261 @@
+
+        public override DiagnosticAnalyzerCategory GetAnalyzerCategory()
+            => DiagnosticAnalyzerCategory.SemanticSpanAnalysis;
+
+        protected override void InitializeWorker(AnalysisContext context)
+            => context.RegisterSyntaxNodeAction(AnalyzeSyntax, this.IfStatementKind);
+
+        private void AnalyzeSyntax(SyntaxNodeAnalysisContext context)
+        {
+            var cancellationToken = context.CancellationToken;
+            var ifStatement = (TIfStatementSyntax)context.Node;
+            var semanticModel = context.SemanticModel;
+
+            var option = context.GetAnalyzerOptions().PreferCoalesceExpression;
+            if (!option.Value)
+                return;
+
+            var syntaxFacts = this.SyntaxFacts;
+            var condition = GetConditionOfIfStatement(ifStatement);
+
+            if (!IsNullCheck(condition, out var checkedExpression))
+                return;
+
+            var previousStatement = TryGetPreviousStatement(ifStatement);
+            if (previousStatement is null)
+                return;
+
+            if (HasElseBlock(ifStatement))
+                return;
+
+            if (!TryGetEmbeddedStatement(ifStatement, out var whenTrueStatement))
+                return;
+
+            if (syntaxFacts.IsThrowStatement(whenTrueStatement))
+            {
+                if (!syntaxFacts.SupportsThrowExpression(ifStatement.SyntaxTree.Options))
+                    return;
+
+                var thrownExpression = syntaxFacts.GetExpressionOfThrowStatement(whenTrueStatement);
+                if (thrownExpression is null)
+                    return;
+            }
+
+            TExpressionSyntax? expressionToCoalesce;
+
+            if (syntaxFacts.IsLocalDeclarationStatement(previousStatement))
+            {
+                // var v = Expr();
+                // if (v == null)
+                //    ...
+
+                if (!AnalyzeLocalDeclarationForm(previousStatement, out expressionToCoalesce))
+                    return;
+            }
+            else if (syntaxFacts.IsAnyAssignmentStatement(previousStatement))
+            {
+                // v = Expr();
+                // if (v == null)
+                //    ...
+                if (!AnalyzeAssignmentForm(previousStatement, out expressionToCoalesce))
+                    return;
+            }
+            else
+            {
+                return;
+            }
+
+            context.ReportDiagnostic(DiagnosticHelper.Create(
+                Descriptor,
+                ifStatement.GetFirstToken().GetLocation(),
+                option.Notification.Severity,
+                ImmutableArray.Create(
+                    expressionToCoalesce.GetLocation(),
+                    ifStatement.GetLocation(),
+                    whenTrueStatement.GetLocation()),
+                properties: null));
+
+            bool AnalyzeLocalDeclarationForm(
```

Comment:
idk if there's a time when local methods are too big and should be pulled out, but this feels like it.